File: internal/adapter/modelref.go
package adapter

import (
	"fmt"
	"strings"
)

// DefaultAdapterKind is the adapter kind assumed when a model_ref has no
// explicit adapter segment (a 1- or 2-segment legacy ref). "opencode" is
// the only adapter wired into the dispatcher today; legacy refs like
// "opencode/deepseek-v4-flash-free" and "anthropic/claude-sonnet-4" keep
// dispatching exactly as they did before the adapter namespace landed.
const DefaultAdapterKind = "opencode"

// ModelRef is the parsed form of a worker model_ref under the pinned
// left-greedy grammar (ADR-0003):
//
//	<adapter>/<provider>/<model>   3+ segments (explicit adapter kind)
//	<provider>/<model>             2 segments (legacy — adapter inferred)
//	<model>                        1 segment (legacy — adapter inferred)
//
// The model field is the remainder after the first two segments, VERBATIM
// (internal slashes preserved — e.g. the Command Code ref
// "orchicon/commandcode/deepseek/deepseek-v4-flash" parses to
// Model "deepseek/deepseek-v4-flash"). The parser NEVER splits the model
// field at an internal slash, and model-id validation must not forbid "/".
type ModelRef struct {
	Adapter  string // segment 1 (3+ seg) or the inferred legacy default
	Provider string // segment 2 (3+ seg) or segment 1 (legacy 2-seg)
	Model    string // verbatim remainder (internal slashes preserved)
}

// ParseModelRef parses a model ref under the pinned grammar (left-greedy,
// exactly three semantic fields).
//
//   - "adapter/provider/model" and deeper ("adapter/provider/a/b/c") parse
//     left-greedy with the model = verbatim remainder.
//   - A legacy 2-segment "provider/model" ref infers the adapter from the
//     default adapter kind when seg1 is a KNOWN provider of that adapter
//     (built-in profile OR tenant-created custom provider via reg). A
//     2-segment ref whose seg1 is a known adapter kind is rejected as
//     malformed (it should be written adapter/provider/model); an unknown
//     seg1 is rejected with a Settings → Adapters pointer.
//   - A 1-segment ref is a legacy bare model id; adapter defaults.
//
// reg may be nil (e.g. deep in the adapter layer where no tenant registry
// is available): validation of known providers/adapters is skipped and the
// ref is parsed structurally. All errors are actionable (ADR-0003 D4).
func ParseModelRef(ref string, reg ProviderRegistry) (ModelRef, error) {
	ref = strings.TrimSpace(ref)
	if ref == "" {
		return ModelRef{}, fmt.Errorf("model ref must not be empty (expected adapter/provider/model or the legacy provider/model)")
	}
	i := strings.IndexByte(ref, '/')
	hasSlash := i >= 0
	seg1, rest := splitFirst(ref)
	seg1 = strings.TrimSpace(seg1)
	if seg1 == "" {
		return ModelRef{}, fmt.Errorf("model ref %q must be adapter/provider/model (or the legacy provider/model)", ref)
	}
	if !hasSlash {
		// 1-segment legacy: a bare model id, adapter defaults.
		return ModelRef{Adapter: DefaultAdapterKind, Model: seg1}, nil
	}
	seg2, modelRest := splitFirst(rest)
	seg2 = strings.TrimSpace(seg2)
	if seg2 == "" {
		return ModelRef{}, fmt.Errorf("model ref %q must be adapter/provider/model (or the legacy provider/model)", ref)
	}
	if modelRest == "" {
		// 2-segment legacy: provider/model with the adapter inferred.
		if reg != nil {
			if reg.IsKnownProvider(DefaultAdapterKind, seg1) {
				return ModelRef{Adapter: DefaultAdapterKind, Provider: seg1, Model: seg2}, nil
			}
			if reg.IsKnownAdapter(seg1) {
				return ModelRef{}, fmt.Errorf("model ref %q: %q is an adapter kind, not a provider — use adapter/provider/model (e.g. %s/provider/model)", ref, seg1, seg1)
			}
			return ModelRef{}, fmt.Errorf("unknown provider %q in model ref %q — create it in Settings → Adapters (custom provider), or use adapter/provider/model", seg1, ref)
		}
		return ModelRef{Adapter: DefaultAdapterKind, Provider: seg1, Model: seg2}, nil
	}
	// 3+ segments: left-greedy. Segment 1 = adapter, segment 2 = provider,
	// the remainder = model VERBATIM (internal slashes preserved — the
	// pinned fix for slashed model ids like Command Code's).
	if reg != nil && !reg.IsKnownAdapter(seg1) {
		return ModelRef{}, fmt.Errorf("unknown adapter %q in model ref %q — register an adapter of that kind, or use adapter/provider/model with a known adapter (e.g. %s/provider/model)", seg1, ref, DefaultAdapterKind)
	}
	return ModelRef{Adapter: seg1, Provider: seg2, Model: modelRest}, nil
}

// AdapterKind extracts the adapter kind from a model ref WITHOUT
// validation — the routing view used by the dispatcher/reconciler (an
// unknown kind surfaces its actionable "register an adapter" error at
// Resolve). Left-greedy: segment 1 of a 3+ segment ref; DefaultAdapterKind
// for a 1- or 2-segment legacy ref; "" for an empty/malformed ref.
func AdapterKind(ref string) string {
	ref = strings.TrimSpace(ref)
	if ref == "" {
		return ""
	}
	seg1, rest := splitFirst(ref)
	seg1 = strings.TrimSpace(seg1)
	if seg1 == "" {
		return ""
	}
	i := strings.IndexByte(ref, '/')
	hasSlash := i >= 0
	if !hasSlash {
		// 1-segment legacy: a bare model id, adapter defaults.
		return DefaultAdapterKind
	}
	if strings.TrimSpace(rest) == "" {
		// Trailing slash — malformed; the caller resolves that.
		return ""
	}
	seg2, _ := splitFirst(rest)
	if strings.TrimSpace(seg2) == "" {
		return ""
	}
	// 3+ segments: the adapter kind is segment 1. A 2-segment ref (no
	// further slash in rest) is legacy: adapter defaults to opencode.
	if strings.IndexByte(rest, '/') < 0 {
		return DefaultAdapterKind
	}
	return seg1
}

// SplitForServe splits a model ref into the provider/model pair the serve
// API expects (opencode serve still receives {"providerID","modelID"};
// the adapter segment is consumed by the control plane and dropped before
// the serve call — ADR-0003 D1/D6). Left-greedy over the first two
// segments: a 3+ segment ref uses segment 2 as the provider and the
// verbatim remainder as the model; a legacy 2-segment ref uses segment 1
// as the provider and segment 2 as the model. ok=false when no
// provider/model can be derived (empty or 1-segment ref).
//
// The 3+-segment branch assumes CANONICAL refs (adapter/provider/model).
// Pre-namespace legacy refs with a slashed head (e.g. "commandcode/a/b")
// are NOT canonical: their head is a legacy provider id, not an adapter
// kind, and left-greedy would yield provider=seg2 — the WRONG serve pair.
// The canonicalize-pre-namespace-refs migration (20260917000000) rewrites
// such rows into canonical form, and worker validation's phantom-kind rule
// (internal/worker/validate.go) rejects new saves of that shape, so this
// path only ever sees canonical 3+-segment refs in practice. It is kept
// purely structural (no registry access) because serve/usage callers deep
// in the adapter layer have no tenant registry.
func SplitForServe(ref string) (provider, model string, ok bool) {
	ref = strings.TrimSpace(ref)
	if ref == "" {
		return "", "", false
	}
	seg1, rest := splitFirst(ref)
	seg1 = strings.TrimSpace(seg1)
	if rest == "" || seg1 == "" {
		return "", "", false
	}
	seg2, modelRest := splitFirst(rest)
	seg2 = strings.TrimSpace(seg2)
	if seg2 == "" {
		return "", "", false
	}
	if modelRest == "" {
		// 2-segment legacy: provider/model.
		return seg1, seg2, true
	}
	// 3+ segments: provider = segment 2, model = verbatim remainder.
	return seg2, modelRest, true
}

// splitFirst splits ref at its first "/", returning the head and the
// remainder ("" when there is no separator).
func splitFirst(ref string) (head, rest string) {
	i := strings.IndexByte(ref, '/')
	if i < 0 {
		return ref, ""
	}
	return ref[:i], ref[i+1:]
}

// NormalizeRef rewrites a PRE-NAMESPACE legacy model ref into canonical
// adapter/provider/model form. Legacy refs are free-form provider/<model>
// entries written before the adapter namespace existed (ADR-0003): their
// head can be a plain provider id (opencode-go), a slashed model path
// (commandcode/deepseek/deepseek-v4-flash), or a bare model id. The new
// left-greedy grammar misreads such refs — head becomes an "adapter kind",
// which poisons adapter-change validation (a phantom adapter selection
// nobody made), dispatch row-kind resolution, and SplitForServe's serve
// pair. NormalizeRef converts them to the canonical shape:
//
//	opencode-go/deepseek-v4-flash   → opencode/opencode-go/deepseek-v4-flash
//	commandcode/a/b/c               → opencode/commandcode/a/b/c
//	bare-model                      → opencode/bare-model  (bare ids need no rewrite)
//	opencode/anthropic/m            → unchanged (already canonical)
//
// knownAdapters decides the legacy-vs-canonical cut: a head that IS a
// registered adapter kind means the ref already carries an adapter segment
// and is returned unchanged; anything else is treated as legacy and
// normalized under DefaultAdapterKind. Pass nil to normalize structurally
// (every 1- and 2-segment ref, plus ANY 3+ segment ref whose head is not
// a known kind — used by the data migration where the dispatcher's live
// kinds are unavailable). Refs are otherwise returned VERBATIM (no
// re-splitting of model segments, no trimming beyond the stored value).
func NormalizeRef(ref string, knownKinds map[string]struct{}) string {
	ref = strings.TrimSpace(ref)
	if ref == "" {
		return ""
	}
	first, rest := splitFirst(ref)
	first = strings.TrimSpace(first)
	if first == "" {
		return ref // malformed (empty head); leave for parse-time errors
	}
	if knownKinds != nil {
		if _, ok := knownKinds[first]; ok {
			return ref // already carries a registered adapter segment
		}
	}
	// No further slash → 2-segment legacy (provider/model) or 1-segment
	// bare id. Both normalize by prefixing the default adapter kind.
	if rest == "" || !strings.Contains(rest, "/") {
		return DefaultAdapterKind + "/" + ref
	}
	// 3+ segments with a non-kind head: legacy slashed provider ref.
	return DefaultAdapterKind + "/" + ref
}

// NormalizeRefForMigration is NormalizeRef for the data migration: it
// reports (normalized, changed) — changed=true only when the normalized
// form differs from the input, so the migration skips canonical and
// empty/malformed rows untouched.
func NormalizeRefForMigration(ref string, knownKinds map[string]struct{}) (string, bool) {
	out := NormalizeRef(ref, knownKinds)
	if out == "" {
		return ref, false
	}
	return out, out != ref
}

File: frontend/src/lib/model-ref.ts
// model-ref.ts - TypeScript mirror of the Go model_ref grammar
// (internal/adapter/modelref.go, ADR-0003).
//
// The grammar is left-greedy over exactly three semantic fields:
//
//   <adapter>/<provider>/<model>   3+ segments (explicit adapter kind)
//   <provider>/<model>             2 segments (legacy - adapter inferred as opencode)
//   <model>                        1 segment (legacy - adapter inferred as opencode)
//
// The model field is the remainder after the first two segments, VERBATIM
// (internal slashes preserved - e.g. "orchicon/commandcode/deepseek/deepseek-v4-flash"
// parses to Model "deepseek/deepseek-v4-flash"). The parser NEVER splits the
// model field at an internal slash.

/** The adapter kind assumed for legacy refs with no explicit adapter segment. */
export const DEFAULT_ADAPTER_KIND = "opencode";

/**
 * ORCHICON_ADAPTER_KIND is the decided product default for FRESH adapter
 * selections (ADR-0005 D5): the native adapter kind this feature delivers.
 * The picker seeds its adapter tier with this kind when the stored ref is
 * empty AND the kind is Dispatcher-registered; otherwise it degrades to
 * DEFAULT_ADAPTER_KIND (opencode) — never a default that cannot dispatch.
 * It is deliberately NOT the backward-compat inference: legacy refs
 * without an adapter segment keep inferring opencode.
 */
export const ORCHICON_ADAPTER_KIND = "orchicon";

/** The parsed form of a model_ref. */
export interface ParsedModelRef {
  adapter: string; // segment 1 (3+ seg) or the inferred legacy default
  provider: string; // segment 2 (3+ seg) or segment 1 (legacy 2-seg); "" for 1-seg
  model: string; // verbatim remainder (internal slashes preserved)
  /** Raw ref verbatim, so unknown/malformed refs keep rendering flagged. */
  raw: string;
}

/**
 * parseModelRef parses a model ref under the pinned left-greedy grammar.
 * Structural parse only - no catalog validation (the frontend has no
 * registry; unknown adapter/provider/model is the picker's flagged-review
 * state, D5). Returns null for empty/malformed refs (a raw ref with no
 * slash and no segments, or empty segments).
 */
export function parseModelRef(ref: string | undefined | null): ParsedModelRef | null {
  const trimmed = (ref ?? "").trim();
  if (!trimmed) return null;

  const [seg1, ...rest] = trimmed.split("/");
  if (!seg1 || seg1.trim() === "") return null;

  if (rest.length === 0) {
    // 1-segment legacy: a bare model id, adapter defaults.
    return { adapter: DEFAULT_ADAPTER_KIND, provider: "", model: seg1, raw: trimmed };
  }
  const seg2 = rest[0];
  if (!seg2 || seg2.trim() === "") return null;

  if (rest.length === 1) {
    // 2-segment legacy: provider/model, adapter inferred.
    return { adapter: DEFAULT_ADAPTER_KIND, provider: seg1, model: seg2, raw: trimmed };
  }

  // 3+ segments: left-greedy - model is the verbatim remainder.
  return {
    adapter: seg1,
    provider: seg2,
    model: rest.slice(1).join("/"),
    raw: trimmed,
  };
}

/**
 * CatalogModelLike is the minimal catalog-entry shape model-match needs
 * (structural — no protobuf type import). `modelRef` is the catalog's
 * LEGACY 2-segment "providerId/id" (internal/aigateway OpenCodeModel);
 * `providerId`+`id` carry the authoritative segments.
 */
export interface CatalogModelLike {
  id: string;
  providerId: string;
  modelRef?: string; // legacy 2-segment "providerId/id"
}

/**
 * catalogModelMatches reports whether a catalog entry corresponds to the
 * model in a (parsed) model ref, by SEGMENTS — not by raw-string equality.
 * A catalog entry's modelRef is the legacy 2-segment "providerId/id", so
 * a raw comparison against a 3-segment ref ("opencode/anthropic/claude-4")
 * never matches; the adapter segment is inherently not part of the
 * catalog's identity (models are scoped per provider+id). A parsed ref is
 * REQUIRED: unknown shapes (null) deliberately never match, so callers
 * keep flagging them for review (ADR-0004 D5).
 */
export function catalogModelMatches(
  parsed: ParsedModelRef,
  model: CatalogModelLike,
): boolean {
  if (parsed.provider !== "") {
    if (parsed.provider === model.providerId && parsed.model === model.id) return true;
    // Fallback for entries with empty segments: compare against the legacy
    // 2-segment modelRef verbatim.
    return model.modelRef === `${parsed.provider}/${parsed.model}`;
  }
  // Legacy 1-segment ref: a bare model id with no provider segment.
  return parsed.model === model.id || parsed.model === model.modelRef;
}

/**
 * formatModelRef builds the canonical 3-segment ref "adapter/provider/model".
 * The model segment is written verbatim (may contain internal slashes).
 * Empty provider/model collapse the ref back toward legacy forms only when
 * absent - callers pass a full selection; the function itself never
 * fabricates segments.
 */
export function formatModelRef(adapter: string, provider: string, model: string): string {
  const a = adapter.trim();
  const p = provider.trim();
  const m = model.trim();
  if (!a) return m;
  if (!p) return m ? `${a}/${m}` : a;
  return m ? `${a}/${p}/${m}` : `${a}/${p}`;
}
